src/Streamly/Internal/FileSystem/Event/Linux.hs

```diff
@@ -216,10 +225,170 @@ setFlag mask status cfg@Config{..} =
                 Off -> createFlags .&. complement mask
     in cfg {createFlags = flags}
 
+-- | Mapping of recursive subdirectories with root dir.
+-- key : subdir path 
+-- value : tuple of root path and relative subdir path  
+-- /Internal/
+--
+type RootMap = IORef (M.Map FilePath (FilePath, FilePath))
+
+-- | Exclude dir . and .. 
+-- while generating list of subdir recursively 
+-- /Internal/
+--
+exclude :: FilePath -> Bool
+exclude path = head path /= '.'
+
+-- | Traverse a dir recusively and return the list of subdirs.
+--
+-- /Internal/
+--
+traverseDirRec :: 
+       FilePath -- ^ subdir path
+    -> FilePath -- ^ root dir path
+    -> [FilePath] -- ^ roots passed on commnd line
+    -> StateT ([FilePath], RootMap) IO [FilePath] -- ^ Mapping with root dir
```

Comment:
Use the code from the `ListDir.hs` example to return a stream of sudirs here. Then use `S.foldlM'` on the resulting stream to apply the function that adds each dir to the watch handle and return the accumulated result. The accumulator can contain the map or any other aggregated value that you may want to return. We do not need a `StateT`.